client/app/scripts/components/sparkline.js

```
@@ -1,118 +1,179 @@
+  }
+
+  componentWillMount() {
+    this.setState(this.updateSlidingWindow(this.props));
+  }
+
+  componentWillUnmount() {
+    clearTimeout(this.tickTimer);
+  }
+
+  componentWillReceiveProps(nextProps) {
+    this.setState(this.updateSlidingWindow(nextProps));
+  }
+
+  componentDidUpdate() {
+    // move sliding window one tick
+    if (!this.tickTimer) {
+      this.tick();
+    }
   }
 
-  renderSparkline() {
-    // If the sparkline has already been rendered, remove it.
-    const el = ReactDOM.findDOMNode(this);
-    while (el.firstChild) {
-      el.removeChild(el.firstChild);
+  tick() {
+    if (this.state.buffer.size > 0 && this.state.last < this.state.buffer.keySeq().last()) {
+      const dates = this.state.buffer.keySeq();
+      let firstIndex = dates.indexOf(this.state.first);
+      if (firstIndex > -1 && firstIndex < dates.size - 1) {
+        firstIndex++;
+      } else {
+        firstIndex = 0;
+      }
+      const first = dates.get(firstIndex);
+
+      let lastIndex = dates.indexOf(this.state.last);
+      if (lastIndex > -1 && lastIndex < dates.size) {
+        lastIndex++;
+      } else {
+        lastIndex = dates.length - 1;
+      }
+      const last = dates.get(lastIndex);
+
+      this.tickTimer = setTimeout(() => {
+        this.tickTimer = null;
+        this.setState({first, last});
+      }, 900);
     }
+  }
 
-    const data = this.props.data.slice();
+  getDataFromBuffer() {
+    const dateFilter = d => d.date >= this.state.first && d.date <= this.state.last;
+    return this.state.buffer.map((v, k) => {
+      return {value: v, date: k};
+    }).toIndexedSeq().toJS().filter(dateFilter);
+  }
+
+  getGraphData() {
+    // data is of shape [{date, value}, ...]
+    let data = this.getDataFromBuffer();
 
-    // Do nothing if no data is passed in.
-    if (data.length === 0) {
-      return;
+    // Do nothing if no data or data w/o date are passed in.
+    if (data.length === 0 || data[0].date === undefined) {
```

fail fast? +1 for exploding if we're getting some weird data.